### PR TITLE
Add Command SYSTEM FLUSH SETTINGS DISTRIBUTED

### PR DESCRIPTION
**Pull request by sunny19930321**
### Changelog category (leave one):
- New Feature

this PR fix issue https://github.com/ClickHouse/ClickHouse/issues/41924

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

SYSTEM FLUSH SETTINGS DISTRIBUTED distributed_table;

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

File: src/Storages/Distributed/DistributedAsyncInsertDirectoryQueue.cpp [hunk 1]
```diff
@@ -427,22 +438,41 @@ void DistributedAsyncInsertDirectoryQueue::processFile(const std::string & file_
...
+            LOG_DEBUG(
+                log,
+                "Sending `{}` to {} ({} rows, {} bytes)",
+                file_path,
+                connection->getDescription(),
+                formatReadableQuantity(distributed_header.rows),
+                formatReadableSizeWithBinarySuffix(distributed_header.bytes));
+
+            RemoteInserter remote{
+                *connection, timeouts, distributed_header.insert_query, distributed_header.insert_settings, distributed_header.client_info};
```
Comment (Enmk) [inline src/Storages/Distributed/DistributedAsyncInsertDirectoryQueue.cpp:471]: If the only difference betwee two blocks is the only variable, maybe it is worth to factor that variable out?

Like that:
```
const Settings & settings_for_insert = is_flush_settings ? context_->getSettingsRef() : distributed_header.insert_settings;```

And just change everything else almost intact?

File: src/Storages/StorageDistributed.cpp [hunk 2]
```diff
@@ -1797,6 +1797,26 @@ void StorageDistributed::flushClusterNodesAllData(ContextPtr local_context)
         node->flushAllData();
 }
 
+void StorageDistributed::flushClusterNodesAllSettings(ContextPtr local_context)
```
Comment (Enmk) [inline src/Storages/StorageDistributed.cpp:1800]: This is almost 100% copy of `StorageDistributed::flushClusterNodesAllData`, I suggest factoring out populating `directory_monitors` into a separate function and re-using it.

File: src/Storages/Distributed/DistributedAsyncInsertDirectoryQueue.cpp (hunk 1 above)
Comment (sunny19930321) [reply src/Storages/Distributed/DistributedAsyncInsertDirectoryQueue.cpp]: ok

File: src/Storages/StorageDistributed.cpp (hunk 2 above)
Comment (sunny19930321) [reply src/Storages/StorageDistributed.cpp:1807]: ok

File: tests/integration/test_distributed_system_flush_settings_query/test.py [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+    node1.query("SET max_partitions_per_insert_block = 1;")
+
+    node1.query("INSERT INTO distributed_table VALUES (0, 'node1'),(1, 'node2')")
+
+    # Write only to this node when stop distributed sends
+    assert node1.query("SELECT COUNT() FROM distributed_table").rstrip() == "1"
+
+    node1.query("SET max_partitions_per_insert_block = 0;")
+    node1.query("SYSTEM START DISTRIBUTED SENDS distributed_table;")
+    node1.query("SYSTEM FLUSH SETTINGS DISTRIBUTED distributed_table;")
```
Comment (azat) [inline tests/integration/test_distributed_system_flush_settings_query/test.py:42]: Actually I think it should be `SYSTEM FLUSH DISTRIBUTED SENDS db.table SETTINGS ...`
In other words simply add `SETTINGS` clause support for `SYSTEM FLUSH DISTRIBUTED SENDS`

**Comment by azat** [comment]
>Run Check — Category 'New Feature Add Command' is not valid

Looks like a typo in the changelog entry

>Style Check — Check black failed.

Worth to run `black` for your new test.

File: tests/integration/test_distributed_system_flush_settings_query/test.py (hunk 3 above)
Comment (sunny19930321) [reply tests/integration/test_distributed_system_flush_settings_query/test.py:42]: > Actually I think it should be `SYSTEM FLUSH DISTRIBUTED SENDS db.table SETTINGS ...` In other words simply add `SETTINGS` clause support for `SYSTEM FLUSH DISTRIBUTED SENDS`

OK

**Comment by robot-clickhouse-ci-1** [comment]
This is an automated comment for commit 21138c00ba91c4057b4cda31681019a4e4b30a52 with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/49089/21138c00ba91c4057b4cda31681019a4e4b30a52/ci_running.html)
The overall status of the commit is 🟡 pending

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🟡 pending</td></tr>
<tr><td>Docs Check</td><td>Builds and tests the documentation</td><td>🟡 pending</td></tr>
</table>

**Comment by alexey-milovidov** [comment]
Tests have to be updated.

**Comment by sunny19930321** [comment]
> Tests have to be updated.

OK

**Comment by alexey-milovidov** [comment]
I understand the purpose of this change, but:
- the feature is not intuitive to use, it is not user-friendly, it is difficult to forget why it is needed;
- this PR has an accidental modification of submodules and some other files.

**Comment by alexey-milovidov** [comment]
The motivation is understood - we can submit another PR. @azat, do you want to submit it?

**Comment by azat** [comment]
Let's try something like this - https://github.com/ClickHouse/ClickHouse/pull/61832

**Pull request closed without merging**